File: ios/NetCap/Sources/NetcapBridge.swift
import Foundation

/// NetcapBridge - iOS bridge for the netcap FFI library.
///
/// This class wraps the UniFFI-generated Swift bindings to provide
/// a clean iOS-friendly API for the netcap proxy functionality.
///
/// Usage:
/// ```swift
/// let bridge = NetcapBridge(port: 8080, storagePath: documentsPath)
/// try bridge.start()
/// let stats = try bridge.getStats()
/// try bridge.stop()
/// ```
public class NetcapBridge {
    private let port: UInt16
    private let storagePath: String
    private let includeDomains: [String]
    private let excludeDomains: [String]
    private var isRunning = false

    public init(
        port: UInt16 = 8080,
        storagePath: String,
        includeDomains: [String] = [],
        excludeDomains: [String] = []
    ) {
        self.port = port
        self.storagePath = storagePath
        self.includeDomains = includeDomains
        self.excludeDomains = excludeDomains
    }

    /// Start the proxy server.
    public func start() throws {
        guard !isRunning else {
            throw NetcapError.alreadyRunning
        }
        // TODO: Call UniFFI-generated NetcapProxy.start()
        isRunning = true
    }

    /// Stop the proxy server.
    public func stop() throws {
        guard isRunning else {
            throw NetcapError.notRunning
        }
        // TODO: Call UniFFI-generated NetcapProxy.stop()
        isRunning = false
    }

    /// Get the CA certificate in PEM format.
    public func getCaCertificatePem() throws -> String {
        // TODO: Call UniFFI-generated NetcapProxy.getCaCertificatePem()
        return ""
    }

    /// Get current capture statistics.
    public func getStats() throws -> CaptureStats {
        // TODO: Call UniFFI-generated NetcapProxy.getStats()
        return CaptureStats(totalRequests: 0, totalResponses: 0, activeConnections: 0, bytesCaptured: 0)
    }

    /// Get captured events as JSON string.
    public func getCaptureEvents(offset: UInt64 = 0, limit: UInt64 = 100) throws -> String {
        // TODO: Call UniFFI-generated NetcapProxy.getCaptureEvents()
        return "[]"
    }

    public var proxyRunning: Bool { isRunning }
}

/// Capture statistics from the proxy.
public struct CaptureStats {
    public let totalRequests: UInt64
    public let totalResponses: UInt64
    public let activeConnections: UInt32
    public let bytesCaptured: UInt64
}

/// Errors from the netcap proxy.
public enum NetcapError: Error {
    case initFailed(String)
    case proxyError(String)
    case alreadyRunning
    case notRunning
    case storageError(String)
    case certError(String)
}

File: android/app/src/main/kotlin/com/netcap/bridge/NetcapBridge.kt
package com.netcap.bridge

/**
 * NetcapBridge - Android bridge for the netcap FFI library.
 *
 * This class wraps the UniFFI-generated Kotlin bindings to provide
 * a clean Android-friendly API for the netcap proxy functionality.
 *
 * Usage:
 * ```kotlin
 * val bridge = NetcapBridge(port = 8080, storagePath = context.filesDir.absolutePath)
 * bridge.start()
 * val stats = bridge.getStats()
 * bridge.stop()
 * ```
 */
class NetcapBridge(
    private val port: Int = 8080,
    private val storagePath: String,
    private val includeDomains: List<String> = emptyList(),
    private val excludeDomains: List<String> = emptyList()
) {
    private var isRunning = false

    /**
     * Start the proxy server.
     * @throws IllegalStateException if the proxy is already running
     */
    fun start() {
        if (isRunning) {
            throw IllegalStateException("Proxy is already running")
        }
        // TODO: Call UniFFI-generated NetcapProxy.start()
        isRunning = true
    }

    /**
     * Stop the proxy server.
     * @throws IllegalStateException if the proxy is not running
     */
    fun stop() {
        if (!isRunning) {
            throw IllegalStateException("Proxy is not running")
        }
        // TODO: Call UniFFI-generated NetcapProxy.stop()
        isRunning = false
    }

    /**
     * Get the CA certificate in PEM format.
     * This certificate needs to be installed on the device to intercept HTTPS traffic.
     */
    fun getCaCertificatePem(): String {
        // TODO: Call UniFFI-generated NetcapProxy.getCaCertificatePem()
        return ""
    }

    /**
     * Get current capture statistics.
     */
    fun getStats(): CaptureStats {
        // TODO: Call UniFFI-generated NetcapProxy.getStats()
        return CaptureStats(0, 0, 0, 0)
    }

    /**
     * Get captured events as JSON string.
     */
    fun getCaptureEvents(offset: Long = 0, limit: Long = 100): String {
        // TODO: Call UniFFI-generated NetcapProxy.getCaptureEvents()
        return "[]"
    }

    fun isProxyRunning(): Boolean = isRunning
}

/**
 * Data class representing capture statistics.
 */
data class CaptureStats(
    val totalRequests: Long,
    val totalResponses: Long,
    val activeConnections: Int,
    val bytesCaptured: Long
)
